pep-0723.rst

```diff
@@ -0,0 +1,185 @@
+PEP: 723
+Title: Embedding metadata in single-file scripts
+Author: Ofek Lev <ofekmeister@gmail.com>
+PEP-Delegate: Brett Cannon <brett@python.org>
+Discussions-To: https://discuss.python.org/t/29905
+Status: Draft
+Type: Standards Track
+Topic: Packaging
+Content-Type: text/x-rst
+Created: 04-Aug-2023
+Post-History: `19-Jul-2023 <https://discuss.python.org/t/29905>`__
+
+
+Abstract
+========
+
+This PEP specifies a format for defining metadata about single-file Python
+scripts, such as their runtime dependencies. This PEP merely extends
+:pep:`722` with an alternative format and therefore elides describing the
+motivation for such a proposal.
+
+Specification
+=============
+
+This PEP chooses to follow the latest developments of other modern packaging
+ecosystems (namely `Rust <https://github.com/rust-lang/rfcs/blob/master/text/3424-cargo-script.md>`__
+and `Go <https://github.com/erning/gorun>`__) by embedding the existing
+standard :pep:`621` that is used to describe projects.
+
+Any Python script may assign a variable named ``__pyproject__`` to a multi-line
+double-quoted string containing a valid TOML document. This document contains
+what you would expect to see in a ``pyproject.toml`` file, with the following
+exceptions regarding the ``[project]`` table:
+
+* The ``name`` and ``version`` fields MUST be defined dynamically by tools if
+  the user does not define them
+* These fields do not need to be listed in the ``dynamic`` array
+
+Example
+-------
+
+The following is an example of a script with an embedded ``pyproject.toml``:
+
+.. code:: python
+
+    __pyproject__ = """
+    [project]
+    requires-python = ">=3.11"
+    dependencies = [
+      "requests<3",
+      "rich",
+    ]
+    """
+
+    import requests
+    from rich.pretty import pprint
+
+    resp = requests.get("https://peps.python.org/api/peps.json")
+    data = resp.json()
+    pprint([(k, v["title"]) for k, v in data.items()][:10])
+
+
+Reference Implementation
+========================
+
+This regular expression may be used to parse the metadata::
+
+    (?ms)^__pyproject__ *= *"""$(.+?)^"""$
+
```

Comment:
Is this regular expression the specification of the format, or just an example? Because it does not handle the full range of possibilities here. What about

```
__pyproject__ = """\
    [project]
    ...
"""
```

While the backslash isn't necessary here, I regularly use multi-line strings in a context where a leading newline would be an error, so I tend to add the backslash without thinking. So I'd be surprised to find that a spec that says "a multi-line double-quoted string containing a valid TOML document" disallows this possibility.

There's also the possibility of people using doubled backslashes to include a backslash in the TOML.

This is fundamentally the issue with "use Python notation". Are you using Python, or are you using some subset? If the latter, then you need to define it explicitly, or *someone* will insist that the standard says you must support some obscure facet of Python syntax that you didn't even consider.

Reply:
Probably, I should have posted what I wrote in the [discourse thread](https://discuss.python.org/t/pep-723-embedding-metadata-in-single-file-scripts/30979/6?u=abravalheri) directly here.

The following are examples (non-exhaustive) of syntaxes that (to my understanding, and I might be wrong) seem to be permitted by the current state of the draft, are valid Python code, result in a string value (after Python's syntax interpretation) that is valid TOML, but might break the reference implementation:

```python
__pyproject__ = """
[project]
readme.text = \"""
Hello, this is my awesome single-file utility script.
Try running::

    pipx run myfile.py
\"""
dependencies = ['tomli; python_version < "3.11"']
"""

__pyproject__ = """
project.dependencies = """ """\
['tomli; python_version < "3.11"']
"""

__pyproject__ = (
   """\
   project.dependencies = ['tomli; python_version < "3.11"']
   """
)  # Auto-formatters sometimes add parenthesis...
```

Reply:
I added:

> In circumstances where there is a discrepancy between the regular expression and the text specification, the text specification takes precedence.

Reply:
Thanks, but in which case the text specification isn't clear enough (IMO) to write a parser for it. Note that I'm thinking about this from the perspective of potentially having to implement this for pipx, *not* as the author of a competing PEP.

And to be clear, if the PEP says that there is the potential for a discrepancy between the suggested regex and the formal spec, I'd personally avoid using the regex, because I prefer correctness over convenience. So this is a real issue for me.